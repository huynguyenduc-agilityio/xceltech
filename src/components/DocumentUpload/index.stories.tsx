import { Meta, StoryObj } from '@storybook/react';

// Components
import DocumentUpload from '.';

const meta = {
  title: 'Components/DocumentUpload',
  component: DocumentUpload,
  parameters: {
    layout: 'centered',
  },
  render: (arg) => (
    <div className="flex justify-center w-[1000px] p-9 bg-white">
      <DocumentUpload {...arg} />
    </div>
  ),
  tags: ['autodocs'],
} satisfies Meta<typeof DocumentUpload>;

export default meta;

type Story = StoryObj<typeof DocumentUpload>;

export const Default: Story = {
  args: {
    onFileChange: () => {},
  },
};
