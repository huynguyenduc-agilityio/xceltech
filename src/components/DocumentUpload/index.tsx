import { useCallback, useEffect, useState } from 'react';

import { Button, Input, Label } from '../common';

export interface DocumentUploadProps {
  value?: File | string | null;
  onFileChange: (file: File) => void;
}

const DocumentUpload = ({ value, onFileChange }: DocumentUploadProps) => {
  const [fileUrl, setFileUrl] = useState<string | null>(null);

  const handleFileChange = useCallback(
    async (e: React.ChangeEvent<HTMLInputElement>) => {
      const file = e.target.files?.[0];

      if (file) {
        onFileChange(file);
      }
    },
    [onFileChange],
  );

  useEffect(() => {
    if (value instanceof File) {
      const url = URL.createObjectURL(value);

      setFileUrl(url);

      return () => {
        URL.revokeObjectURL(url);
      };
    }

    if (typeof value === 'string') {
      setFileUrl(value);
    }

    if (!value) {
      setFileUrl(null);
    }
  }, [value]);

  const fileName =
    typeof value === 'string'
      ? (value.split('/').pop() ?? '')
      : value instanceof File
        ? value.name
        : '';

  return (
    <div className="relative flex items-center h-[54px]">
      <Button
        asChild
        className="absolute inset z-10 bg-black-shadow hover:bg-black-shadow w-[254px] py-[12px] text-white h-full rounded"
      >
        <Label htmlFor="file-upload" className="cursor-pointer">
          Choose File
        </Label>
      </Button>
      <Input
        id="file-upload"
        type="file"
        className="hidden"
        onChange={handleFileChange}
        accept=".pdf,.jpg,.png,.docx"
      />
      <div className="absolute flex items-center px-6 w-full h-full bg-blue-light rounded-regular">
        <span className="ml-[254px] text-md truncate text-black-smoky-">
          {fileUrl ? (
            <a href={fileUrl} target="_blank" rel="noopener noreferrer">
              {fileName}
            </a>
          ) : (
            ''
          )}
        </span>
      </div>
    </div>
  );
};

export default DocumentUpload;
