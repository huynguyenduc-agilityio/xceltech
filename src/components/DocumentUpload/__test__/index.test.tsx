import { fireEvent, render } from '@testing-library/react';
import DocumentUpload from '..';

describe('DocumentUpload component', () => {
  const mockOnFileChange = jest.fn();

  beforeAll(() => {
    global.URL.createObjectURL = jest.fn(() => 'mocked-url');
    global.URL.revokeObjectURL = jest.fn();
  });

  afterAll(() => {
    (global.URL.createObjectURL as jest.Mock).mockRestore?.();
    (global.URL.revokeObjectURL as jest.Mock).mockRestore?.();
  });

  it('should match snapshot', () => {
    const { container } = render(
      <DocumentUpload onFileChange={mockOnFileChange} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('renders correctly and allows file selection', () => {
    const testFile = new File(['dummy content'], 'testfile.pdf', {
      type: 'application/pdf',
    });

    const { getByText, getByLabelText } = render(
      <DocumentUpload value={null} onFileChange={mockOnFileChange} />,
    );

    const label = getByText('Choose File');
    expect(label).toBeInTheDocument();

    const input = getByLabelText('Choose File') as HTMLInputElement;
    fireEvent.change(input, { target: { files: [testFile] } });

    expect(mockOnFileChange).toHaveBeenCalledWith(testFile);
  });

  it('displays the selected file name', () => {
    const file = new File(['data'], 'selected-file.png', { type: 'image/png' });

    const { getByText } = render(
      <DocumentUpload value={file} onFileChange={jest.fn()} />,
    );

    expect(getByText('selected-file.png')).toBeInTheDocument();
  });
});
