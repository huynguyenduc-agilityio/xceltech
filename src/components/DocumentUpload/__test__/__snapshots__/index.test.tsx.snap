// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentUpload component should match snapshot 1`] = `
<div>
  <div
    class="relative flex items-center h-[54px]"
  >
    <label
      class="peer-disabled:cursor-not-allowed peer-disabled:opacity-70 font-productsans inline-flex items-center justify-center whitespace-nowrap text-lg font-bold ring-offset-background transition-all duration-300 ease-in-out outline-none focus-visible:outline-none disabled:opacity-50 disabled:cursor-not-allowed absolute inset z-10 bg-black-shadow hover:bg-black-shadow w-[254px] py-[12px] text-white h-full rounded cursor-pointer"
      for="file-upload"
    >
      Choose File
    </label>
    <input
      accept=".pdf,.jpg,.png,.docx"
      autocomplete="off"
      class="font-productsans w-full px-5 py-4 text-lg text-black-smoky placeholder:text-gray-soft outline-none disabled:cursor-not-allowed disabled:opacity-50 bg-blue-light h-[54px] hidden"
      data-testid="input-element"
      id="file-upload"
      type="file"
    />
    <div
      class="absolute flex items-center px-6 w-full h-full bg-blue-light rounded-regular"
    >
      <span
        class="ml-[254px] text-md truncate text-black-smoky-"
      />
    </div>
  </div>
</div>
`;
